### PR TITLE
Detect Successful Start in LongPolling and SSE

### PR DESCRIPTION
**Pull request by mikaelm12**
So this is hacky and doesn't work yet but I wanted to get some feedback.
Currently storing a first request flag in the HttpContext which is probably 👎 

Basically wanted some help with the LongPolling transport code. Not setting the ContentLength or the Status code if flush has already been called.

### CODE REVIEW

File: src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs [hunk 1]
```diff
@@ -30,10 +30,18 @@ public LongPollingTransport(CancellationToken timeoutToken, PipeReader applicati
         {
             try
             {
+                var headersFlushed = false;
+                if (context.Items.TryGetValue("FirstRequest", out var firstRequest))
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs]: Why can't you just pass it into the constructor? The transport itself isn't long lived.

File: src/Microsoft.AspNetCore.Sockets.Client.Http/ServerSentEventsTransport.cs [hunk 2]
```diff
@@ -68,17 +69,24 @@ public Task StartAsync(Uri url, IDuplexPipe application, TransferMode requestedT
...
-        private async Task OpenConnection(IDuplexPipe application, Uri url, CancellationToken cancellationToken)
+        private async Task OpenConnection(IDuplexPipe application, Uri url, TaskCompletionSource<object> tcs, CancellationToken cancellationToken)
         {
             Log.StartReceive(_logger);
 
             var request = new HttpRequestMessage(HttpMethod.Get, url);
             SendUtils.PrepareHttpRequest(request, _httpOptions);
             request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("text/event-stream"));
             var response = await _httpClient.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, cancellationToken);
+            if (!response.IsSuccessStatusCode)
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/ServerSentEventsTransport.cs]: Why not use `response.EnsureSuccessStatusCode()` like the long polling transport does?

Comment (anurse) [reply src/Microsoft.AspNetCore.Sockets.Client.Http/ServerSentEventsTransport.cs]: You want the exception that is thrown by `EnsureSuccessStatusCode` to manifest on the TCS.

Having said that, I actually think `EnsureSuccessStatusCode` coupled with a `try`-`catch` that puts the exception into the TCS (like in Long Polling) would be good as it would capture the stack properly.

Comment (anurse) [reply src/Microsoft.AspNetCore.Sockets.Client.Http/ServerSentEventsTransport.cs]: Also, we need the `try`-`catch` anyway since not all errors will manifest as a failed response, some will manifest as exceptions thrown by `SendAsync` (socket errors such as not being able to reach the server will manifest that way)

File: src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs (hunk 1 above)
Comment (anurse) [reply src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs]: Yeah, that works. We were dealing with the restriction of the `IHttpTransport` interface, but you're right, the ctor is invoked for each incoming poll.

File: src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs [hunk 3]
```diff
@@ -30,10 +30,18 @@ public LongPollingTransport(CancellationToken timeoutToken, PipeReader applicati
...
+                    if ((bool)firstRequest) {
+                        await context.Response.Body.FlushAsync();
+                        headersFlushed = true;
+                    }
+                }
                 var result = await _application.ReadAsync(token);
                 var buffer = result.Buffer;
 
-                if (buffer.IsEmpty && result.IsCompleted)
+                if (buffer.IsEmpty && result.IsCompleted && !headersFlushed)
```
Comment (anurse) [inline src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs]: Put a comment here saying that if we get an empty&completed result in the first request, we'll have to send a 200 because we've already flushed the headers, but that's OK since the client will poll again and immediately get the 204.

File: src/Microsoft.AspNetCore.Sockets.Http/HttpConnectionDispatcher.cs [hunk 4]
```diff
@@ -192,6 +192,11 @@ public HttpConnectionDispatcher(ConnectionManager manager, ILoggerFactory logger
                             Log.PollCanceled(_logger, connection.ConnectionId, existing.TraceIdentifier);
                         }
                     }
+                    else
+                    {
+                        // Hack to indicate that we're handling the first poll.
+                        context.Items.Add("FirstRequest", true);
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.Sockets.Http/HttpConnectionDispatcher.cs]: I can't comment on whether this is a good idea but `"FirstRequest"` should be a const field somewhere

File: src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs (hunk 1 above)
Comment (JamesNK) [inline src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs]: Use const

File: test/Microsoft.AspNetCore.Sockets.Tests/HttpConnectionDispatcherTests.cs [hunk 5]
```diff
@@ -499,6 +499,7 @@ public HttpConnectionDispatcherTests(ITestOutputHelper output) : base(output)
         {
             using (StartLog(out var loggerFactory, LogLevel.Debug))
             {
+                //while (!System.Diagnostics.Debugger.IsAttached) { }
```
Comment (JamesNK) [inline test/Microsoft.AspNetCore.Sockets.Tests/HttpConnectionDispatcherTests.cs]: remove

File: test/Microsoft.AspNetCore.Sockets.Tests/HttpConnectionDispatcherTests.cs [hunk 6]
```diff
@@ -753,6 +757,8 @@ public HttpConnectionDispatcherTests(ITestOutputHelper output) : base(output)
...
```
Comment (JamesNK) [inline test/Microsoft.AspNetCore.Sockets.Tests/HttpConnectionDispatcherTests.cs]: remove

File: test/Microsoft.AspNetCore.Sockets.Tests/HttpConnectionDispatcherTests.cs [hunk 7]
```diff
@@ -773,9 +779,18 @@ public HttpConnectionDispatcherTests(ITestOutputHelper output) : base(output)
                 // Write to the application
                 await connection.Application.Output.WriteAsync(buffer);
 
-                await task;
+                // The fir
```
Comment (JamesNK) [inline test/Microsoft.AspNetCore.Sockets.Tests/HttpConnectionDispatcherTests.cs]: remove

**Comment by mikaelm12** [comment]
Still very WIP. Working through more tests

File: src/Microsoft.AspNetCore.Sockets.Http/HttpConnectionDispatcher.cs (hunk 4 above)
Comment (anurse) [reply src/Microsoft.AspNetCore.Sockets.Http/HttpConnectionDispatcher.cs]: We're changing it to a constructor argument to LongPollingTransport (since it's created on each request), which means no more magic strings.

File: src/Microsoft.AspNetCore.Sockets.Client.Http/LongPollingTransport.cs [hunk 8]
```diff
@@ -85,7 +86,7 @@ public Task StartAsync(Uri url, IDuplexPipe application, TransferMode requestedT
             }
         }
 
-        private async Task Poll(Uri pollUrl, CancellationToken cancellationToken)
+        private async Task Poll(Uri pollUrl, TaskCompletionSource<object> startTcs,CancellationToken cancellationToken)
```
Comment (anurse) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/LongPollingTransport.cs]: nit: spacing after comma

Finally! My turn to remark on spacing :P

File: src/Microsoft.AspNetCore.Sockets.Client.Http/ServerSentEventsTransport.cs [hunk 9]
```diff
@@ -80,6 +81,18 @@ public Task StartAsync(Uri url, IDuplexPipe application, TransferMode requestedT
             request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("text/event-stream"));
             var response = await _httpClient.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, cancellationToken);
 
+            try
```
Comment (anurse) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/ServerSentEventsTransport.cs]: Move the `try` up to encompass the `SendAsync` too. That will throw if there's a socket error.

File: src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs [hunk 10]
```diff
@@ -17,23 +17,31 @@ public class LongPollingTransport : IHttpTransport
...
+            _isFirstRequest = isFirstRequest;
         }
 
         public async Task ProcessRequestAsync(HttpContext context, CancellationToken token)
         {
             try
             {
+                if (_isFirstRequest)
+                {
+                        await context.Response.Body.FlushAsync();
```
Comment (anurse) [inline src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs]: nit: indent

File: test/Microsoft.AspNetCore.Sockets.Tests/HttpConnectionDispatcherTests.cs [hunk 11]
```diff
@@ -513,6 +514,9 @@ public HttpConnectionDispatcherTests(ITestOutputHelper output) : base(output)
                 var app = builder.Build();
                 await dispatcher.ExecuteAsync(context, new HttpSocketOptions(), app);
 
+                Assert.Equal(StatusCodes.Status200OK, context.Response.StatusCode);
+
+                await dispatcher.ExecuteAsync(context, new HttpSocketOptions(), app);
```
Comment (anurse) [inline test/Microsoft.AspNetCore.Sockets.Tests/HttpConnectionDispatcherTests.cs]: Put a comment explaining why two requests are needed

File: src/Microsoft.AspNetCore.Sockets.Client.Http/LongPollingTransport.cs [hunk 12]
```diff
@@ -135,6 +137,7 @@ public Task StartAsync(Uri url, IDuplexPipe application, TransferMode requestedT
             }
             catch (Exception ex)
             {
+                startTcs.TrySetException(ex);
                 Log.ErrorPolling(_logger, pollUrl, ex);
                 throw;
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/LongPollingTransport.cs]: Make sure a unit test hits this catch

File: src/Microsoft.AspNetCore.Sockets.Client.Http/ServerSentEventsTransport.cs [hunk 13]
```diff
@@ -80,6 +81,18 @@ public Task StartAsync(Uri url, IDuplexPipe application, TransferMode requestedT
...
+            {
+                response.EnsureSuccessStatusCode();
+                tcs.SetResult(null);
+            }
+            catch (Exception ex)
+            {
+                Log.TransportStopping(_logger);
+                tcs.TrySetException(ex);
+                return;
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/ServerSentEventsTransport.cs]: Make sure a unit test hits this catch

File: src/Microsoft.AspNetCore.Sockets.Client.Http/LongPollingTransport.cs [hunk 14]
```diff
@@ -121,18 +123,24 @@ public Task StartAsync(Uri url, IDuplexPipe application, TransferFormat transfer
                         Log.ReceivedMessages(_logger);
 
                         var stream = new PipeWriterStream(_application.Output);
-                        await response.Content.CopyToAsync(stream);
+                        try
+                        {
+                            await _response.Content.CopyToAsync(stream);
+                        }
+                        catch { /*First request*/};
```
Comment (anurse) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/LongPollingTransport.cs]: Why would this throw? If it's an expected condition, can we check it before calling `CopyToAsync`?

File: src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs [hunk 15]
```diff
@@ -73,20 +92,22 @@ public LongPollingTransport(CancellationToken timeoutToken, PipeReader applicati
...
                     context.Response.ContentType = "text/plain";
                     context.Response.StatusCode = StatusCodes.Status200OK;
                 }
                 else
                 {
                     // Case 3
-                    Log.LongPolling204(_logger);
-                    context.Response.ContentType = "text/plain";
-                    context.Response.StatusCode = StatusCodes.Status204NoContent;
+                    if (!_isFirstRequest)
```
Comment (anurse) [inline src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs]: You can merge this with the `else` above into an `else if`

File: test/Microsoft.AspNetCore.SignalR.Client.Tests/HttpConnectionTests.ConnectionLifecycle.cs [hunk 16]
```diff
@@ -237,17 +237,7 @@ Task OnTransportStart()
...
-                    var longPollResult = new TaskCompletionSource<HttpResponseMessage>();
-                    httpHandler.OnLongPoll(cancellationToken =>
-                    {
-                        cancellationToken.Register(() =>
-                        {
-                            longPollResult.TrySetResult(ResponseUtils.CreateResponse(HttpStatusCode.NoContent));
-                        });
-                        return longPollResult.Task;
-                    });
+                    var httpHandler = TestHttpMessageHandler.CreateDefault(handleSend:false);
```
Comment (anurse) [inline test/Microsoft.AspNetCore.SignalR.Client.Tests/HttpConnectionTests.ConnectionLifecycle.cs]: spacing

File: test/Microsoft.AspNetCore.SignalR.Client.Tests/HttpConnectionTests.SendAsync.cs [hunk 17]
```diff
@@ -96,16 +95,7 @@ public Task SendThrowsIfConnectionIsDisposed()
...
-
-                var longPollTcs = new TaskCompletionSource<HttpResponseMessage>(TaskCreationOptions.RunContinuationsAsynchronously);
-
-                testHttpHandler.OnLongPoll(cancellationToken =>
-                {
-                    cancellationToken.Register(() => longPollTcs.TrySetResult(null));
-
-                    return longPollTcs.Task;
-                });
+                var testHttpHandler = TestHttpMessageHandler.CreateDefault(handleSend:false);
```
Comment (anurse) [inline test/Microsoft.AspNetCore.SignalR.Client.Tests/HttpConnectionTests.SendAsync.cs]: Dr. Spaceman!

File: test/Microsoft.AspNetCore.SignalR.Client.Tests/TestHttpMessageHandler.cs [hunk 18]
```diff
@@ -10,6 +10,7 @@ namespace Microsoft.AspNetCore.SignalR.Client.Tests
     public class TestHttpMessageHandler : HttpMessageHandler
     {
         private Func<HttpRequestMessage, CancellationToken, Task<HttpResponseMessage>> _handler;
+        public bool IsFirstRequest = true;
```
Comment (anurse) [inline test/Microsoft.AspNetCore.SignalR.Client.Tests/TestHttpMessageHandler.cs]: `IsFirstLongPoll` or something. There are various kinds of requests handled by this handler :)

File: test/Microsoft.AspNetCore.SignalR.Client.Tests/TestHttpMessageHandler.cs [hunk 19]
```diff
@@ -28,20 +29,28 @@ public TestHttpMessageHandler(bool autoNegotiate = true)
...
+                    // Just block until canceled
+                    var tcs = new TaskCompletionSource<object>();
+                    using (cancellationToken.Register(() => tcs.TrySetResult(null)))
+                    {
+                        await tcs.Task;
+                    }
+                    return ResponseUtils.CreateResponse(HttpStatusCode.NoContent);
                 }
-                return ResponseUtils.CreateResponse(HttpStatusCode.NoContent);
+                testHttpMessageHandler.IsFirstRequest = false;
```
Comment (anurse) [inline test/Microsoft.AspNetCore.SignalR.Client.Tests/TestHttpMessageHandler.cs]: Comment: "Otherwise, if this is the first request, just toggle the first request flag and send an empty response"

File: test/Microsoft.AspNetCore.Sockets.Tests/HttpConnectionDispatcherTests.cs [hunk 20]
```diff
@@ -512,8 +512,13 @@ public HttpConnectionDispatcherTests(ITestOutputHelper output) : base(output)
                 var builder = new ConnectionBuilder(services.BuildServiceProvider());
                 builder.UseEndPoint<ImmediatelyCompleteEndPoint>();
                 var app = builder.Build();
+
+                // We need two requests because te first just establishes the transport.
```
Comment (anurse) [inline test/Microsoft.AspNetCore.Sockets.Tests/HttpConnectionDispatcherTests.cs]: nit: `the`

File: src/Microsoft.AspNetCore.Sockets.Client.Http/LongPollingTransport.cs (hunk 14 above)
Comment (mikaelm12) [reply src/Microsoft.AspNetCore.Sockets.Client.Http/LongPollingTransport.cs]: This call to copytoasync was throwing on the first request. We should be keeping track of the first request on the client side transport as well and add that check here.

File: src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs [hunk 21]
```diff
@@ -17,38 +17,66 @@ public class LongPollingTransport : IHttpTransport
...
+                {
+                    await context.Response.Body.FlushAsync();
+                }
+
+                ReadResult result;
+                try
+                {
+                    result = await _application.ReadAsync(token);
+                }
+                catch (Exception) when (_isFirstRequest)
```
Comment (mikaelm12) [inline src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs]: We can't send a 500 on the first request since the headers have already been sent so we catch here. The next poll will throw

**Comment by mikaelm12** [comment]
This is ready for review

File: src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs (hunk 21 above)
Comment (anurse) [reply src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs]: Put a note in the comment below about how the next poll should throw the same exception, and the client will get the error in that poll.

File: src/Microsoft.AspNetCore.Sockets.Client.Http/LongPollingTransport.cs [hunk 22]
```diff
@@ -23,7 +23,7 @@ public partial class LongPollingTransport : ITransport
         private IDuplexPipe _application;
         private Task _sender;
         private Task _poller;
-
+        private HttpResponseMessage _response;
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/LongPollingTransport.cs]: Why is this being stored?

Comment (mikaelm12) [reply src/Microsoft.AspNetCore.Sockets.Client.Http/LongPollingTransport.cs]: Because we needed to dispose the response in StopAsync. This is irrelevant now though as per Andrew's suggestion above

File: src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs (hunk 21 above)
Comment (davidfowl) [reply src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs]: What's the point in reading after flushing?

File: src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs [hunk 23]
```diff
@@ -17,38 +17,67 @@ public class LongPollingTransport : IHttpTransport
         private readonly ILogger _logger;
         private readonly CancellationToken _timeoutToken;
         private readonly string _connectionId;
+        private readonly bool _isFirstRequest;
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs]: This entire change should be reverted and we should just flush from the outside on first request and avoid calling into the long polling transport altogether.

Comment (mikaelm12) [reply src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs]: awesome

Comment (anurse) [reply src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs]: Wait, so just flush a blank `200` for the first request? The change as it stands still writes the first message to the first request if there is one. The only time we have to flush a blank `200` is when there's no data in the pipe. It would make things simpler but all Long Polling connections would require two requests to initialize.

File: src/Microsoft.AspNetCore.Sockets.Client.Http/LongPollingTransport.cs [hunk 24]
```diff
@@ -92,12 +92,11 @@ public Task StartAsync(Uri url, IDuplexPipe application, TransferFormat transfer
...
                     var request = new HttpRequestMessage(HttpMethod.Get, pollUrl);
                     SendUtils.PrepareHttpRequest(request, _httpOptions);
-
-                    HttpResponseMessage response;
+                    HttpResponseMessage _response;
 
                     try
                     {
-                        response = await _httpClient.SendAsync(request, cancellationToken);
+                        _response = await _httpClient.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, cancellationToken);
```
Comment (anurse) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/LongPollingTransport.cs]: You don't declare `_response` any more :)

File: src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs (hunk 23 above)
Comment (davidfowl) [reply src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs]: Ah I misunderstood this change a bit. Still I hate the complexity of the control flow here. Can we expose 2 methods instead of passing the bool?

Comment (anurse) [reply src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs]: Not sure I see how that would help, it would have to either a) Call into a single method with a bool (thus defeating the purpose) or b) Duplicate quite a bit of logic.

File: src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs (hunk 21 above)
Comment (anurse) [reply src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs]: As mentioned earlier: because we still try to write out the first batch of messages in this request, we just do it via chunked encoding.

**Comment by mikaelm12** [comment]
Reverted the changes to `ProcessRequestAsync` and move the first request logic to `ProcessFirstRequestAsync` as per @davidfowl's request

File: src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs [hunk 25]
```diff
@@ -58,26 +89,21 @@ public LongPollingTransport(CancellationToken timeoutToken, PipeReader applicati
                     // Complete the ReadAsync by indicating we haven't processed any data yet
                     // (of course, there is no data because the buffer is empty, but we have to 
                     // call AdvanceTo after every ReadAsync even if we don't have any data)
-                    _application.AdvanceTo(buffer.Start);
+                    //_application.AdvanceTo(buffer.Start);
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs]: ? Remove dead code?

File: src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs [hunk 26]
```diff
@@ -18,23 +18,79 @@ public class LongPollingTransport : IHttpTransport
         private readonly CancellationToken _timeoutToken;
         private readonly string _connectionId;
 
-        public LongPollingTransport(CancellationToken timeoutToken, PipeReader application, string connectionId, ILoggerFactory loggerFactory)
+        public LongPollingTransport(CancellationToken timeoutToken, PipeReader application, string connectionId, ILoggerFactory loggerFactory, bool isFirstRequest = false)
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs]: Remove the  bool isFirstRequest argument.

File: src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs [hunk 27]
```diff
@@ -26,15 +26,70 @@ public LongPollingTransport(CancellationToken timeoutToken, PipeReader applicati
             _logger = loggerFactory.CreateLogger<LongPollingTransport>();
         }
 
+        public async Task ProcessFirstRequestAsync(HttpContext context, CancellationToken token)
+        {
+            if (!token.IsCancellationRequested)
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs]: Why are we checking the token here?

Comment (davidfowl) [reply src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs]: If we do decide to keep it, here's a style thing.

if (token.IsCancellationRequested)
{
    return;
}

Comment (mikaelm12) [reply src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs]: I thought we needed it for one of the test edge cases but turns out I can remove it from the method completely

File: src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs [hunk 28]
```diff
@@ -26,15 +26,70 @@ public LongPollingTransport(CancellationToken timeoutToken, PipeReader applicati
...
+                try
+                {
+                    await context.Response.Body.FlushAsync();
+
+                    ReadResult result;
+                    try
+                    {
+                        result = await _application.ReadAsync(token);
+                    }
+                    catch (Exception)
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs]: Remove the  (Exception) we aren't doing anything with it. catch will suffice.

File: src/Microsoft.AspNetCore.Sockets.Client.Http/LongPollingTransport.cs [hunk 29]
```diff
@@ -53,7 +52,8 @@ public Task StartAsync(Uri url, IDuplexPipe application, TransferFormat transfer
             Log.StartTransport(_logger, transferFormat);
 
             // Start sending and polling (ask for binary if the server supports it)
-            _poller = Poll(url, _transportCts.Token);
+            var startTcs = new TaskCompletionSource<object>();
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/LongPollingTransport.cs]: Ah almost forgot this, this should look like:

```C#
new TaskCompletionSource<int>(TaskCreationOptions.RunContinuationsAsynchronously)
```

We don't want to run user code on our threads.

File: src/Microsoft.AspNetCore.Sockets.Client.Http/ServerSentEventsTransport.cs [hunk 30]
```diff
@@ -50,12 +50,13 @@ public Task StartAsync(Uri url, IDuplexPipe application, TransferFormat transfer
                 throw new ArgumentException($"The '{transferFormat}' transfer format is not supported by this transport.", nameof(transferFormat));
             }
 
+            var startTcs = new TaskCompletionSource<object>();
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/ServerSentEventsTransport.cs]: Same here see https://github.com/aspnet/SignalR/pull/1566/files#r175983362

File: src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs [hunk 31]
```diff
@@ -26,15 +26,67 @@ public LongPollingTransport(CancellationToken timeoutToken, PipeReader applicati
...
                 var buffer = result.Buffer;
 
+                // IF the buffer is empty and the read result is completed on the first request
+                // we still have to send a 200 because the headers have already been flushed.
                 if (buffer.IsEmpty && result.IsCompleted)
                 {
+                    // Complete the ReadAsync by indicating we haven't processed any data yet
+                    // (of course, there is no data because the buffer is empty, but we have to 
+                    // call AdvanceTo after every ReadAsync even if we don't have any data)
+                    //_application.AdvanceTo(buffer.Start);
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs]: Why is this commented out?

Comment (mikaelm12) [reply src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs]: We needed this before when we were handling the first request in this method

Comment (mikaelm12) [reply src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs]: Actually, I think we still need this..

Comment (davidfowl) [reply src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs]: Why?

Comment (mikaelm12) [reply src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs]: Technically we don't need it because in this branch when the 204 is returned the connection would get disposed. https://github.com/aspnet/SignalR/blob/dev/src/Microsoft.AspNetCore.Sockets.Http/HttpConnectionDispatcher.cs#L250-L253
I was correct before when I said that we needed it when we handled the first request in this method. If we didn't call advance an exception would be thrown. Unnecessary now

File: src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs [hunk 32]
```diff
@@ -26,13 +26,60 @@ public LongPollingTransport(CancellationToken timeoutToken, PipeReader applicati
...
+
         public async Task ProcessRequestAsync(HttpContext context, CancellationToken token)
         {
             try
             {
                 var result = await _application.ReadAsync(token);
                 var buffer = result.Buffer;
 
+                // IF the buffer is empty and the read result is completed on the first request
+                // we still have to send a 200 because the headers have already been flushed.
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs]: This comment is stale.

File: src/Microsoft.AspNetCore.Sockets.Client.Http/LongPollingTransport.cs [hunk 33]
```diff
@@ -121,18 +121,23 @@ public Task StartAsync(Uri url, IDuplexPipe application, TransferFormat transfer
                         Log.ReceivedMessages(_logger);
 
                         var stream = new PipeWriterStream(_application.Output);
-                        await response.Content.CopyToAsync(stream);
+                        using (cancellationToken.Register(res => { ((HttpResponseMessage)res).Dispose(); }, response))
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/LongPollingTransport.cs]: nit: Remove surrounding braces

File: test/Microsoft.AspNetCore.SignalR.Client.Tests/TestHttpMessageHandler.cs [hunk 34]
```diff
@@ -10,6 +10,7 @@ namespace Microsoft.AspNetCore.SignalR.Client.Tests
     public class TestHttpMessageHandler : HttpMessageHandler
     {
         private Func<HttpRequestMessage, CancellationToken, Task<HttpResponseMessage>> _handler;
+        public bool IsFirstLongPoll = true;
```
Comment (davidfowl) [inline test/Microsoft.AspNetCore.SignalR.Client.Tests/TestHttpMessageHandler.cs]: nit: property?

Comment (davidfowl) [reply test/Microsoft.AspNetCore.SignalR.Client.Tests/TestHttpMessageHandler.cs]: Why is this public?

File: src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs [hunk 35]
```diff
@@ -26,13 +26,60 @@ public LongPollingTransport(CancellationToken timeoutToken, PipeReader applicati
             _logger = loggerFactory.CreateLogger<LongPollingTransport>();
         }
 
+        public async Task ProcessFirstRequestAsync(HttpContext context, CancellationToken token)
```
Comment (anurse) [inline src/Microsoft.AspNetCore.Sockets.Http/Internal/Transports/LongPollingTransport.cs]: When we talked yesterday, the plan was to just remove all the Long Polling logic (set both Client and Server back to their behavior before this PR). Long Polling will be dumb, and the client will just assume `StartAsync` works as before (and thus the server doesn't need to do any flushing). This PR is still relevant for the Server-Sent Events changes, but the Long Polling changes can just be reverted entirely
